finality-aleph/src/network/manager/session.rs

```diff
@@ -48,19 +50,38 @@ pub enum HandlerError {
     MultiplePeerIds,
 }
 
-fn retrieve_peer_id(addresses: &[Multiaddr]) -> Result<PeerId, HandlerError> {
+fn get_common_peer_id<M: Multiaddress>(addresses: &[M]) -> Option<M::PeerId> {
+    addresses
+        .iter()
+        .try_fold(None, |maybe_peer_id, address| match address.get_peer_id() {
```

Comment:
why not just `fold` doesn't look like anything can fail here? You could then get rid of `flatten` too

Reply:
Maybe you could try extracting this closure into its own function? It's certainly complex enough :smile:

Reply:
Changed into something more similar to what was there before the refactor, although unfortunately this resulted in some code duplication. :/

At this moment I'm a tad sad we don't write in a more functional language, there this kind of thing would be solved with completely generic functions. :<